include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+   * @param start is the first buffer byte to copy.
+   * @param length is the length of sequence of buffer bytes to copy.
+   * @param ptr_ptr is the location in the VM address space to place the address of the newly
+   * allocated memory block which contains the copied bytes.
+   * @param size_ptr is the location in the VM address space to place the size of the newly
+   * allocated memory block which contains the copied bytes (e.g. length).
+   * @return true on success.
+   */
+  virtual bool copyTo(WasmBase *wasm, size_t start, size_t length, uint64_t ptr_ptr,
+                      uint64_t size_ptr) const = 0;
+};
+
+/**
+ * PluginBase is container to hold plugin information which is shared with all Context(s) created
+ * for a given plugin. Embedders may extend this class with additional host-specific plugin
+ * information as required.
+ */
+struct PluginBase {
+  PluginBase(string_view name, string_view root_id, string_view vm_id,
+             string_view plugin_configuration)
+      : name_(std::string(name)), root_id_(std::string(root_id)), vm_id_(std::string(vm_id)),
+        plugin_configuration_(plugin_configuration) {}
+
+  const std::string name_;
+  const std::string root_id_;
+  const std::string vm_id_;
+  std::string plugin_configuration_;
+  const std::string &log_prefix() const { return log_prefix_; }
+
+private:
+  std::string makeLogPrefix() const;
+
+  std::string log_prefix_;
+};
+
+/**
+ * ContextBase is the interface between the VM host and the VM. It has several uses:
+ *
+ * 1) To provide host-specific implementations of ABI calls out of the VM. For example, a proxy
+ * which wants to provide the ability to make an HTTP call must implement the
+ * ContextBase::httpCall() method.
+ *
+ * 2) To call into the VM. For example, when the above mentioned httpCall() completes, the host must
+ * call ContextBase::onHttpCallResponse(). Similarly, when a new HTTP request arrives and the
+ * headers are available, the host must create a new ContextBase object to manage the new stream and
+ * call onRequestHeaders() on that object which will cause a corresponding Context to be allocated
+ * in the VM which will receive the proxy_on_context_create and proxy_on_request_headers calls.
+ *
+ * 3) For testing and instrumentation the methods of ContextBase can be replaces or augmented.
+ */
+class ContextBase {
+public:
+  ContextBase();                                                   // Testing.
+  ContextBase(WasmBase *wasm);                                     // Vm Context.
+  ContextBase(WasmBase *wasm, std::shared_ptr<PluginBase> plugin); // Root Context.
+  ContextBase(WasmBase *wasm, uint32_t root_context_id,
+              std::shared_ptr<PluginBase> plugin); // Stream context.
+  virtual ~ContextBase();
+
+  WasmBase *wasm() const { return wasm_; }
+  uint32_t id() const { return id_; }
+  bool isVmContext() { return id_ == 0; }
+  bool isRootContext() { return root_context_id_ == 0; }
```

Comment:
Can you add some method comments on those methods where it's not super clear what they mean or what zero means? I think that would be helpful for implementors.

Reply:
Done.